File: 103.二叉树的锯齿形层序遍历.cpp
/*
 * @lc app=leetcode.cn id=103 lang=cpp
 *
 * [103] 二叉树的锯齿形层序遍历
 */
#include "algorithm"
#include "iostream"
#include "map"
#include "queue"
#include "vector"

using namespace std;
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
  public:
    vector<vector<int>> zigzagLevelOrder(TreeNode *root) {
        if (!root)
            return {};
        vector<vector<int>> v;
        queue<TreeNode *> q;
        q.push(root);
        bool b = true;
        while (q.size()) {
            vector<int> vv;
            vector<TreeNode *> vvv;
            while (q.size()) {
                TreeNode *node = q.front();
                q.pop();
                vv.push_back(node->val);
                if (node->left) {
                    vvv.push_back(node->left);
                }
                if (node->right) {
                    vvv.push_back(node->right);
                }
            }
            for (auto &i : vvv) {
                q.push(i);
            }
            if (!b)
                reverse(vv.begin(), vv.end());
            v.push_back(vv);
            b = !b;
        }
        return v;
    }
};
// @lc code=end

File: 103.二叉树的锯齿形层序遍历.go
/*
 * @lc app=leetcode.cn id=103 lang=golang
 *
 * [103] 二叉树的锯齿形层序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func reverse(s []int) {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
    s[i], s[j] = s[j], s[i]
	}
}

func zigzagLevelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	var result = [][]int{}
	var queue []*TreeNode
	var line = []*TreeNode{root}
	var b = true
	for len(line) > 0 {
		var ret []int
		for len(line) > 0 {
			node := line[0]
			ret = append(ret, node.Val)
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			line = line[1:]
		}
		line = queue
		queue = nil
		if b {
			reverse(ret)
			result = append(result, ret)
		} else {
			result = append(result, ret)
		}
		b = !b
	}
	return result
}
// @lc code=end

